src/dotty/tools/dotc/core/Skolemization.scala

```diff
@@ -0,0 +1,127 @@
+package dotty.tools.dotc
+package core
+
+import Symbols._, Types._, Contexts._
+import collection.mutable
+
+/** Methods to add and remove skolemtypes.
+ *  
+ *  Skolem types are generated when comparing refinements. 
+ *  Two skolem types are equal if they refer to the same underlying type. 
+ *  To avoid unsoundness, skolem types have to be kept strictly local to the
+ *  comparison, they are not allowed to escape the lifetime of a comparison
+ *  by surviving in a context or in GADT bounds. 
+ */
+trait Skolemization {
```

Comment:
Skolemizationating! It'd be great to use this trait's comment to explain once and for all what skolem types are and what their purposes are. I'd be curious to know exactly how they're related to https://en.wikipedia.org/wiki/Skolem_normal_form too.